processing/src/main/java/org/apache/druid/segment/join/JoinFilterAnalyzer.java

```diff
@@ -0,0 +1,743 @@
+              prefixes,
+              equiconditions,
+              correlationCache
+          );
+          if (!rewritten.isCanPushDown()) {
+            return JoinFilterAnalysis.createNoPushdownFilterAnalysis(orFilter);
+          } else {
+            newFilters.add(rewritten.getPushdownFilter());
+          }
+        } else {
+          return JoinFilterAnalysis.createNoPushdownFilterAnalysis(orFilter);
+        }
+      } else {
+        newFilters.add(filter);
+      }
+    }
+
+    return new JoinFilterAnalysis(
+        true,
+        retainRhs,
+        orFilter,
+        new OrFilter(newFilters),
+        null
+    );
+  }
+
+  /**
+   * Rewrites a selector filter on a join table into an IN filter on the base table.
+   *
+   * @param baseAdapter      The adapter for the join
+   * @param selectorFilter   SelectorFilter to be rewritten
+   * @param prefixes         Map of join table prefixes to clauses
+   * @param equiconditions   Map of equiconditions
+   * @param correlationCache Cache of column correlation analyses
+   *
+   * @return A JoinFilterAnalysis that indicates how to handle the potentially rewritten filter
+   */
+  private static JoinFilterAnalysis rewriteSelectorFilter(
+      HashJoinSegmentStorageAdapter baseAdapter,
+      SelectorFilter selectorFilter,
+      Map<String, JoinableClause> prefixes,
+      Map<String, Expr> equiconditions,
+      Map<String, List<JoinFilterColumnCorrelationAnalysis>> correlationCache
+  )
+  {
+    String filteringColumn = selectorFilter.getDimension();
+    for (Map.Entry<String, JoinableClause> prefixAndClause : prefixes.entrySet()) {
+      if (filteringColumn.startsWith(prefixAndClause.getKey())) {
+        List<JoinFilterColumnCorrelationAnalysis> correlations = correlationCache.computeIfAbsent(
+            prefixAndClause.getKey(),
+            p -> findCorrelatedBaseTableColumns(
+                baseAdapter,
+                p,
+                prefixes.get(p),
+                equiconditions
+            )
+        );
+
+        if (correlations == null) {
+          return JoinFilterAnalysis.createNoPushdownFilterAnalysis(selectorFilter);
+        }
+
+        List<Filter> newFilters = new ArrayList<>();
+        List<VirtualColumn> pushdownVirtualColumns = new ArrayList<>();
+
+        for (JoinFilterColumnCorrelationAnalysis correlationAnalysis : correlations) {
+          if (correlationAnalysis.supportsPushDown()) {
+            List<String> correlatedValues = getCorrelatedValuesForPushDown(
+                selectorFilter.getDimension(),
+                selectorFilter.getValue(),
+                correlationAnalysis.getJoinColumn(),
+                prefixAndClause.getValue()
+            );
+
+            if (correlatedValues == null) {
+              return JoinFilterAnalysis.createNoPushdownFilterAnalysis(selectorFilter);
+            }
+
+            for (String correlatedBaseColumn : correlationAnalysis.getBaseColumns()) {
```

Comment:
When would there be more than one base column?

Reply:
Still looking into this comment

Reply:
My original intent was related to the comment here about handling cases where one RHS column is joined to multiple columns: https://github.com/apache/druid/pull/9301#discussion_r375025840

It's not implemented right now in `JoinFilterAnalyzer.findCorrelatedBaseTableColumns` (and should be), but in that case there would be multiple base columns correlated to the RHS filtering column.

Since a query of that type doesn't run successfully right now, can we address this in a follow-on PR?

Reply:
I think doing it in a follow-on is OK. Please just add a comment about the known limitation and create a followup github issue (the comment could just link to the issue).

Reply:
I opened #9327 to track the query failure issue and #9328 to track the adjustments needed here